docs/udf-guide.md

```diff
@@ -0,0 +1,159 @@
+# Guide to User-Defined Functions (UDF)
+
+This is a guide to show how to use UDFs in .NET for Apache Spark.
+
+## What are UDFs
+
+[User-Defined Functions (UDFs)](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/expressions/UserDefinedFunction.html) are a feature of Spark that allow developers to use custom functions to extend the vocabulary of Spark SQL’s Domain Specific Language. They transform values from a single row within a table to produce a single corresponding output value per row based on the logic defined in the UDF.
+
+Let's take the following as an example for a UDF definition:
```

Comment:
I think it'd be great to add some more explanation here (some of which could be before or after the code example), i.e.:

* What does this UDF do?
* Are there multiple ways to define/register UDFs?
* Why do we use "Column" on the left but "string" on the right?
* Some info that the first "string" is the type of the input and the second "string" is the type of the output

Reply:
Have added some explanation and link to relevant sections in our code for easier understanding. Please let me know if this looks fine, thanks!

Reply:
Looks great, minor comment :)